Proposal for updating mip template

andreajrujano opened the pull request.

andreajrujano committed:
feat: WIP suggestions for mip template

andreajrujano committed:
feat: WIP add modules for data bridge and model

andreajrujano commented:
@evpinheiro 
This is a proposal for some changes. That's why is in a separate branch, in this branch, we could implement the proposed version of the Mip template to show it in the Mip stack, or we could wait to discuss it and then implement the changes. So far, it doesn't seem like much work to do.

After some discussion, @LuizSuzana and I considered that at the very least the template "should" have its place for the data bridge which might include or not include data validations, and also for the model. Hence, the main module will orchestrate everything. 

The biggest motivation for this proposal (from my side) is to align it with Mip Way. This is desirable but not mandatory. It's not my main intention to make it more similar Mip Start or merge them.

LuizSuzana commented:
I think we could start developing some things to propose on the meeting, but also because soon we'll need to share this template with the mipsters for them to start implementing their models as part of Mip Way, so it would be good to have the model and data bridge on hand in the template. What do you think @evpinheiro ?

LuizSuzana committed:
feat: commit placeholders for bringing the template closer to Mip Way

LuizSuzana committed:
feat: add TODOs to schemas.py

LuizSuzana committed:
feat: remove uv.lock

LuizSuzana committed:
docs: remove uv instructions from main README

LuizSuzana committed:
feat: add TODOs to actions

LuizSuzana committed:
feat: add back requirements.txt

LuizSuzana committed:
chore: add openpyxl dev dependency

andreajrujano committed:
feat: update schemas

andreajrujano committed:
feat: create model, load_model_data and create_output_data scripts

andreajrujano committed:
feat: add constant for Portions parameter in input schema

andreajrujano committed:
feat: add input action to update cost food

andreajrujano committed:
feat: update main

andreajrujano committed:
chore: update pyproject.toml to add new dependencies

andreajrujano committed:
feat: add imports to __init__.py

andreajrujano committed:
feat: replace action_data_prep by update food cost action

LuizSuzana committed:
refactor: schemas definition, add comment about "parameters" being special

LuizSuzana committed:
feat: remove unnecessary imports in __init__

LuizSuzana committed:
docs: add comment to __main__ about -e argument

LuizSuzana committed:
refactor: update model construction...

1. fix docstring dat -> data_in
2. fix variable name
3. set objective's sense
4. replace pyscipopt's method setRealParam -> setParam, more general
5. print status, and condition output creation by number of feasible solutions
6. return output as a dictionary instead of list

LuizSuzana committed:
feat: update creation of output tables accordingly

LuizSuzana committed:
feat: validate input data in main.py

LuizSuzana committed:
docs: replace mip_template_formulation by the diet problem...

which's a simplified version from mip_start

LuizSuzana committed:
feat: remove unused "Food Portions" schema parameter, modify datatype for the others

LuizSuzana committed:
feat: forbid Min/Max Intake from being null on schemas

LuizSuzana committed:
feat: add Min/Max Intake to output "nutrition" tab

LuizSuzana committed:
feat: print optimal objective value, to examplify getObjVal() method

LuizSuzana committed:
fix: bad imports in main.py

LuizSuzana committed:
feat: update action_update_food_cost to use the actual 'Food Cost Multiplier' param instead of a hard-coded value

LuizSuzana committed:
feat: remove constants.SolverTypes

LuizSuzana committed:
feat: add type hint to path arguments in output action

LuizSuzana committed:
fix: add missing Min/Max Intake columns to output_schema

LuizSuzana committed:
data: add testing_data.xlsx according to input schema

LuizSuzana committed:
feat: include back imports in __init__

LuizSuzana committed:
feat: add foods.Portion column, to allow the user to specify whether a Food ID allows fractional values or not

andreajrujano committed:
feat: update test_local_execution

Changes are:

- Update file data path.
- Change function to be called in the input action to update food cost.
- Add path to save the visualization locally to be passed as argument for `report_builder_solve` function.

andreajrujano committed:
Merge remote-tracking branch 'origin/feat/proposal_update_template' into feat/proposal_update_template

andreajrujano committed:
fix: remove unpacking operator

It was causing to unpack the solution into 3 columns instead of the expected 2.

andreajrujano committed:
fix: add missing column

andreajrujano committed:
feat: implement new output action to create visualization

andreajrujano committed:
feat: add imports to __init__

LuizSuzana committed:
feat: allow foods.Max Intake to be null, skip such constraint in this case

LuizSuzana committed:
refactor: rename portions constants

LuizSuzana committed:
fix: construction of output tables

LuizSuzana committed:
refactor: update method name under test_local_execution.py

LuizSuzana committed:
feat: make the optimization model return more than just variables

LuizSuzana committed:
data: update testing_data.xlsx ...

1. removed foods.Fixed Cost column, not used in this model
2. replaced Food ID values in "foods" and "foods_nutrients" from "n" to "Fn"
3. replaced Nutrient ID values in "nutrients" and "foods_nutrients" from "n" to "Nn"
4. added foods.Portion column, F0 and F2 with value Whole, the others with Fractional
5. formatted foods.Per Unit Cost as dollars
6. formatted as title Food Name and Nutrient Name columns in foods, nutrients, and foods_nutrients
7. formatted nutrients.Min Intake and Max Intake as numbers (thousands and decimals separators)

LuizSuzana committed:
feat: update test_local_execution to read from testing_data.xlsx

LuizSuzana committed:
data: update testing_data.xlsx, remove Food Name and Nutrient Name columns from foods_nutrients

LuizSuzana committed:
chore: add pandas explicit dependency

LuizSuzana committed:
chore: add pyscipopt dependency to requirements.txt

LuizSuzana committed:
fix: KeyError in case model doesn't find any feasible solution

LuizSuzana committed:
fix: only round output tables when storing them into sln, otherwise there is error propagration

LuizSuzana committed:
Revert "data: update testing_data.xlsx, remove Food Name and Nutrient Name columns from foods_nutrients"

This reverts commit ab70400de1e7d1f54f4c2c818c0bd5c5cf44d8ec.

LuizSuzana committed:
feat: add configured deployment to Mip Hub

LuizSuzana committed:
feat: order output tables "buy" and "nutrition" by PK

LuizSuzana committed:
feat: add output "kpis" table

LuizSuzana committed:
fix: update_food_cost_solve action silently overwriting input "dat"

LuizSuzana committed:
refactor: remove hard-coded default value for report_builder_solve

LuizSuzana committed:
feat: update unit tests

LuizSuzana committed:
refactor: rename (test_)mip_template -> (test_)mip_start

LuizSuzana committed:
refactor: remove unused import

LuizSuzana committed:
refactor: rename load_model_data.py -> input_data.py, and create_output_data.py -> output_data.py

andreajrujano committed:
refactor: rename of plot title

LuizSuzana committed:
refactor: rename mip_template -> mip_start in some other places

LuizSuzana committed:
docs: update main README with project setup instructions, including path-handling (to be tested on Windows)

LuizSuzana committed:
docs: remove path-handling instruction for Windows cmd

LuizSuzana committed:
docs: remove suggestions file

LuizSuzana commented:
**Summary of changes**

- implemented the diet problem with a MIP model (using `pyscipopt`)
    - update formulation
    - update pipeline (get optimization data, optimization model, create output tables) to use functions, no classes
    - updated the schema accordingly
    - maintain two actions, input and output, as examples. Output action creates a visualization
- renamed (test_)mip_template -> (test_)mip_start in files and docs
- add instructions to main README for project setup and path-handling
